File: src/main/java/com.foomoo.box/model/immutable/BoardModelDiff.java
package com.foomoo.box.model.immutable;

import com.foomoo.box.Block;
import com.foomoo.box.Cell;
import com.foomoo.box.Target;

import java.util.List;
import java.util.Optional;
import java.util.Set;

import static java.util.stream.Collectors.toList;

/**
 * Class to determine the difference between two {@link BoardModel} objects
 */
public class BoardModelDiff {

    private final BoardModel first;
    private final BoardModel second;

    /**
     * Construct a BoardModelDiff to determine differences between the given BoardModels.
     *
     * @param first  The first model.
     * @param second The second model.
     */
    public BoardModelDiff(final BoardModel first, final BoardModel second) {
        this.first = first;
        this.second = second;
    }

    /**
     * Get the blocks that have changed cell when moving from the first to the second BoardModel.
     *
     * @return The List of Blocks.
     */
    public List<Block> getMovedBlocks() {

        final List<Block> commonBlocks = getCommonBlocks();

        return commonBlocks.stream().filter(block -> !first.getBlockCell(block).equals(second.getBlockCell(block))).collect(toList());
    }

    /**
     * Gets the targets that have been satisfied when moving from the first to the second BoardModel.
     *
     * @return The List of Targets.
     */
    public List<Target> getNewCompletedTargets() {
        final List<Target> firstCompletedTargets = getCompletedTargets(first);
        final List<Target> secondCompletedTargets = getCompletedTargets(second);

        return secondCompletedTargets.stream().filter(secondTarget -> !firstCompletedTargets.contains(secondTarget)).collect(toList());
    }

    /**
     * Gets the targets that have been unsatisfied when moving from the first to the second BoardModel.
     *
     * @return The List of Targets.
     */
    public List<Target> getNewUncompletedTargets() {
        final List<Target> firstCompletedTargets = getCompletedTargets(first);
        final List<Target> secondCompletedTargets = getCompletedTargets(second);

        return firstCompletedTargets.stream().filter(firstTarget -> !secondCompletedTargets.contains(firstTarget)).collect(toList());
    }

    /**
     * Get the Blocks that are present in both the first and second model.
     *
     * @return The List of Blocks.
     */
    private List<Block> getCommonBlocks() {
        final Set<Block> firstBlocks = first.getBlocks();
        final Set<Block> secondBlocks = second.getBlocks();

        return firstBlocks.stream().filter(secondBlocks::contains).collect(toList());
    }

    /**
     * Gets the completed targets for the given BoardModel.
     *
     * @param boardModel The BoardModel to get completed targets for.
     * @return The List of completed Targets.
     */
    private List<Target> getCompletedTargets(final BoardModel boardModel) {
        return boardModel.getTargets().stream().filter(target -> {
            final Optional<Cell> targetCellOptional = boardModel.getTargetCell(target);

            return targetCellOptional.flatMap(targetCell -> boardModel.getBlockAtCell(targetCell).filter(target::isValidBlock)).isPresent();

        }).collect(toList());
    }

}

File: src/main/scala/com/foomoo/box/model/immutable/BoardModelDiff.scala
package com.foomoo.box.model.immutable

import java.util.{Optional, Set}

import com.foomoo.box.{Block, Target}

import scala.collection.JavaConversions._

/**
  * Class to determine the difference between two {@link BoardModel} objects
  */
class BoardModelDiff(first: BoardModel, second: BoardModel) {

  /**
    * Get the blocks that have changed cell when moving from the first to the second BoardModel.
    *
    * @return The List of Blocks.
    */
  def getMovedBlocks: java.util.List[Block] = {
    getCommonBlocks.filter(block => first.getBlockCell(block).get() != second.getBlockCell(block).get()).toList
  }

  /**
    * Gets the targets that have been satisfied when moving from the first to the second BoardModel.
    *
    * @return The List of Targets.
    */
  def getNewCompletedTargets: java.util.List[Target] = (getCompletedTargets(second) &~ getCompletedTargets(first)).toList

  /**
    * Gets the targets that have been unsatisfied when moving from the first to the second BoardModel.
    *
    * @return The List of Targets.
    */
  def getNewUncompletedTargets: java.util.List[Target] = (getCompletedTargets(first) &~ getCompletedTargets(second)).toList

  /**
    * Get the Blocks that are present in both the first and second model.
    *
    * @return The List of Blocks.
    */
  private def getCommonBlocks: Set[Block] = first.getBlocks & second.getBlocks

  /**
    * Gets the completed targets for the given BoardModel.
    *
    * @param boardModel The BoardModel to get completed targets for.
    * @return The List of completed Targets.
    */
  private def getCompletedTargets(boardModel: BoardModel): Set[Target] = {
    boardModel.getTargets.filter(target => {
      toOption(boardModel.getTargetCell(target)) match {
        case None => false
        case Some(targetCell) => {
          toOption(boardModel.getBlockAtCell(targetCell)).map(target.isValidBlock(_)).isDefined
        }
      }
    }
    )
  }

  private def toOption[T](javaOp: Optional[T]): Option[T] = if (javaOp.isPresent) Some(javaOp.get()) else None
}